Create separate pseudo-class invalidation RuleSets for attribute selectors

anttijk opened the pull request:
#### 85f5196bd1f7f680292ee8c4e11754d94180f169
<pre>
Create separate pseudo-class invalidation RuleSets for attribute selectors
<a href="https://bugs.webkit.org/show_bug.cgi?id=297857">https://bugs.webkit.org/show_bug.cgi?id=297857</a>
<a href="https://rdar.apple.com/problem/159102650">rdar://problem/159102650</a>

Reviewed by Matthieu Dubet.

Optimize :has([some-attr=foo]) case.

This helps with GitHub performance as it uses many complex rules of this type.

* Source/WebCore/style/PseudoClassChangeInvalidation.cpp:
(WebCore::Style::makePseudoClassInvalidationKeys):

Look up invalidation rule sets with attribute keys.

* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::unlikelyToHaveSelectorForAttribute):

Ignore id/class/style attributes.

(WebCore::Style::makePseudoClassInvalidationKey):

Looks for attribute selectors and make a separate key for them so attribute selectors don&apos;t end up in
the universal invalidation bucket.

* Source/WebCore/style/RuleFeature.h:

Canonical link: <a href="https://commits.webkit.org/299121@main">https://commits.webkit.org/299121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a734dac0beb9f0bceaef1ed985334bbed9e3c0c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37616 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28250 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124085 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69973 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/77eb9f2d-6ff7-4b34-9d67-15997a3f4d68) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119816 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38309 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46198 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89500 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59109 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/89b1b6de-de21-4363-aa57-599233104aef) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120890 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30506 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105753 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69995 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e25edda5-81ac-47a3-8420-dbdc21d96cc7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29573 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23868 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67748 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99927 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24047 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127166 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44841 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33777 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98167 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45202 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101977 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97955 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24916 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43351 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21332 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41275 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44713 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50387 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44173 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47518 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45862 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/style/RuleFeature.cpp:
```diff
@@ -371,12 +377,18 @@ static PseudoClassInvalidationKey makePseudoClassInvalidationKey(CSSSelector::Ps
     if (!className.isEmpty())
         return makePseudoClassInvalidationKey(pseudoClass, InvalidationKeyType::Class, className);
 
+    if (!attributeName.isEmpty())
```
Is there a reason to favor Class then Attribute then Tag ?

anttijk replied:
The idea is to pick the one most likely be the most specific. I suppose Id -> Attribute -> Class -> Tag might be better, that's the order selector filter uses.

anttijk committed:
Create separate pseudo-class invalidation RuleSets for attribute selectors
https://bugs.webkit.org/show_bug.cgi?id=297857
rdar://problem/159102650

Reviewed by Matthieu Dubet.

Optimize :has([some-attr=foo]) case.

This helps with GitHub performance as it uses many complex rules of this type.

* Source/WebCore/style/PseudoClassChangeInvalidation.cpp:
(WebCore::Style::makePseudoClassInvalidationKeys):

Look up invalidation rule sets with attribute keys.

* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::unlikelyToHaveSelectorForAttribute):

Ignore id/class/style attributes.

(WebCore::Style::makePseudoClassInvalidationKey):

Looks for attribute selectors and make a separate key for them so attribute selectors don't end up in
the universal invalidation bucket.

* Source/WebCore/style/RuleFeature.h:

Canonical link: https://commits.webkit.org/299121@main
